### DISCUSSION TITLE
RBAC Confusion: are API filters enough or do I need custom endpoint for comprehensive access check?

### DISCUSSION THREAD

**Discussion by lumosminima** [pocketbase/pocketbase#6129] [Q&A] [answered]
Greetings👋 I've been building my app without much thought to permissions and was just using simple 'logged in' check from docs but now I have tables like: 

- `users`
- `projects`
  - name
  - projects_roles_ids 
  - invite_token
  - invite_enabled
- `roles`
   - name ~ [ owner, admin, editor, viewer ]
- `permissions`
   - name ~ [ projects.create , projects.delete ... ]
- `roles_permissions` (`roles` <-> `permissions`)
- `projects_roles`
  - project_id
  - user_id
  - roles_permissions_id

A simple-ish RBAC system and a single link invite URL system which could be refreshed or toggled. I'm having trouble thinking in API rules since they act as SQL WHERE and don't know whether I should be using custom endpoint to check access here or are following scenarios covered by API rules.

- User should be able to view all `project_roles` records if they are "owner" of project indicated by roles_permissions_id
   - I only came up with `@request.auth.id ?= user_id && roles_permissions_id.permission_ids.name ?= "roles.view"` and this only returns one record
- I want only "owners" or "admins" to manage roles but what about when a new user needs to join a project and has a valid invite_link. I want to add their role from client side sdk but since I would have API rules in place which take into consideration only existing project members, I won't receive any records hence no way to associate the new user to the project or to the projects_roles. This is my first time handling this and I have some idea how I would handle it in a endpoint but want to know if API rules has some syntax that covers this so I can skip creating a endpoint / hook.

**Comment by ganigeorgiev** [comment]
For me your description is not clear and I personally don't like this type of discussions because there is nothing concrete to answer and usually the only correct reply would be - "it depends", which I imagine doesn't help much.

There is not enough context about your project to understand what is the desired behavior or what the described structure even means in the first place. No one can read your mind or even know what your project is about (_RBAC could mean a lot of things_).

> I only came up with @request.auth.id ?= user_id && role_permission_id.permission_ids.name ?= "roles.view" and this only returns one record

Nowhere in your listed fields there is `permission_ids` - what it is and from where it comes from (is it field in `roles_permission`? what do you mean by `roles`"Singleton"?)?
What are the available records in the related collections and how many records do you expect to be returned (_how can I verify what you are saying is working correctly or not_)?

Regarding your second bulletpoint - what _"has a valid invite_link"_ means (what an "invite" represents? Do the invited users already have a PocketBase account or you want to create it only if they click on some link? How is this link generated? What is `invite_token` and how do you verify it?).

There just too many unknown details to give a definite answer (_or maybe I just have a poor English reading comprehension as it seems other users disagree with me https://github.com/pocketbase/pocketbase/discussions/6079_).

In any case I'll stop engage to avoid further confusion, but you'll have better chance someone to assist you if you:

- Elaborate more on the bigger picture of your project so that the various fields can make sense.

- Be more specific about the exact available fields because any API rule example would depend on your actual collections structure and field names (_you can export your collections configuration as json from the Dashboard > Settings > Export collections_).

- Try to ask more concrete questions and for which exact collections API rule you have troubles making it work (_technically the only question in your post is the title but as mentioned in the beginning - "it depends"_).

**Comment by lumosminima** [comment]
Sorry, I made some typos but I don't know how I could've said it more clearly since I already said I'm "confused". I purposely gave a simplified overview of tables without rest of business specific fields and table _to_ focus on the "bigger picture": typical RBAC flows using API filters and their capabilities compared to SQL or code. This works: `@request.auth.id ?= user_id && roles_permissions_id.permission_ids.name ?= "roles.view"` but only returns one record the "owner" role itself but I want to match/allow all records that have same space_id from that one matched record from the rule. I know how to do it in SQL or code but how to express that in API rules? since they are pocketbase specific thing. Even if you can't read my mind you can have some idea what I wanted to know since RBAC is so common in SaaS apps or atleast could've given me some pointers to look at.

In my past dealing here #5621 you had similar answer and I have since achieved what I wanted there but again I have seen here the common response on different requests is on lines of "pocketbase is purpose built for presentator" so If you can't be bothered guiding on pocketbase specific features like "API rules" or addressing exploratory questions why market the framework as "Open Source backend for Your next SaaS". And yes this counts as marketing since their is time investment. Every SaaS needs are different and if I'm using this "all-in-one" approach I'm bound to run into your design decisions and would want to know best way forward since you made the thing rather than a simple "it depends" with no pros and cons discussed. Isn't that the whole point of framework to provide a narrow path forward? and to not reinvent the wheel because this exactly leads to someone reinventing the wheel. I guess I'm going to stick with what I already know: 'if else' in an endpoint.

**Reply by ganigeorgiev** [reply]
> since RBAC is so common in SaaS apps 

But that was exactly my point. There is no "narrow path forward" because often there is more than one way to accomplish a single thing. RBAC could greatly vary from one use case to another and yours based on the presented structure is definitely not a standard one.

The question is just too vague to answer. While I could to some extend understand your frustration it is equally frustrating for me to give a definite answer.

The rule (assuming that it is for "projects_roles"):

```js
@request.auth.id = user_id && roles_permissions_id.permission_ids.name ?= "roles.view"
```
roughly translates to:

"Returns/Matches all project_roles with user_id equal to the current authenticated user id AND that have at least one roles_permissions_id.permission_ids relation with name equal to 'roles.view'".

If in your case doesn't work as described above, then either there is a bug or I misunderstood your collections structure (_that is why I suggested to export it as json_)


In any case, I'm not forcing anyone to use PocketBase. I only benefit from people using it because of their bug reports and feedback but I don't expect anything in return and I think this is implied already in the [FAQ](https://pocketbase.io/faq/) and my replies in older discussions.

I'm happy to help if I can but trying to guess a solution for a problem when there is no understanding from my part about the said problem, or even about the bigger picture, is beyond what I can or want to do.

If you want a recommendation or suggestion for a problem, please try to ask more concrete questions.

**Reply by ganigeorgiev** [reply]
And just to address the below because it bugs me:

> ... why market the framework as "Open Source backend for Your next SaaS". And yes this counts as marketing since their is time investment.

It is not marketing. I mentioned several times in the past and will repeat it again - **I'm not selling anything or promising anything more than what is already available.** If you consider the website text misleading, please suggest a better one, but behaving like somehow I own you something is wrong and you'll be further disappointed down the road if you approach PocketBase (or any other non-commercial MIT licensed project) with this attitude. Neither you, nor anyone else is entitled to anything. If you disagree, please use something else.

**Reply by lumosminima** [reply]
> > since RBAC is so common in SaaS apps
> 
> But that was exactly my point. There is no "narrow path forward" because often there is more than one way to accomplish a single thing. RBAC could greatly vary from one use case to another and yours based on the presented structure is definitely not a standard one.
> 
> The question is just too vague to answer. While I could to some extend understand your frustration it is equally frustrating for me to give a definite answer.
> 
> The rule (assuming that it is for "projects_roles"):
> 
> ```js
> @request.auth.id = user_id && roles_permissions_id.permission_ids.name ?= "roles.view"
> ```
> 
> roughly translates to:
> 
> "Returns/Matches all project_roles with user_id equal to the current authenticated user id AND that have at least one roles_permissions_id.permission_ids relation with name equal to 'roles.view'".
> 
> If in your case doesn't work as described above, then either there is a bug or I misunderstood your collections structure (_that is why I suggested to export it as json_)
> 
> In any case, I'm not forcing anyone to use PocketBase. I only benefit from people using it because of their bug reports and feedback but I don't expect anything in return and I think this is implied already in the [FAQ](https://pocketbase.io/faq/) and my replies in older discussions.
> 
> I'm happy to help if I can but trying to guess a solution for a problem when there is no understanding from my part about the said problem, or even about the bigger picture, is beyond what I can or want to do.
> 
> If you want a recommendation or suggestion for a problem, please try to ask more concrete questions.

I know there are many ways to do the same things but I'll try to repeat one part of my original question.

The following API Rule returns all records in  "projects_roles" that satisfy the condition as you deducted:

```JS
@request.auth.id = user_id && roles_permissions_id.permission_ids.name ?= "roles.view"
``` 
but since there can only be one "owner" it returns only one record currently which is correct but what I actually want is to return all "projects_roles" records where the "project_id" field is equal to that one record that is currently being matched the "owner". Or to explain how I'm doing it in an endpoint currently:

- Get authenticated user's role (single "projects_roles" record) for a given project_id and store it in a variable x
- If the user's role evaluates to "owner" or has permission "roles.view" (join roles_permissions_id to roles and permissions) 
  -- Get and return all records from "projects_roles" where "project_id" is equal to x.project_id
- Else return only the variable x (user's own "projects_roles" record)

So, does API filter allow referencing some previously matched record so I can reference it's field for a new condition? In the above case I will take "project_id" and then do further matching something that looks in my brain like `(ownerRuleMatchedRecord).project_id ?= project_id` . I don't know how I can explain it more clearly.

**Reply by ganigeorgiev** [reply, marked as answer]
> does API filter allow referencing some previously matched record so I can reference it's field for a new condition?

There is no option to declare and refer to variables in the API rules if that is what you want.

I don't fully understand the desired behavior but I guess what you want is to apply a condition against a relation of the current authenticated user. If that's the case then you should be able to use dot-notation to target the specific relation (either direct or as "indirect", aka. [back-relation](https://pocketbase.io/docs/working-with-relations/#back-relations)). For example:

```js
@request.auth.id != "" &&
// user's own "projects_roles" record
@request.auth.id = user_id || (
  // must have either owner or roles.view permission for the matched project id
  // note: back relations by default resolve as multiple relation field
  @request.auth.projects_roles_via_user_id.project_id ?= project_id && (
    // maybe this should target the "roles" relation of the "roles_permissions" relation?
    // but there is no information what is the name of the relation field so adjust the constraint on your own
    @request.auth.projects_roles_via_user_id.roles_permissions_id.permission_ids.name ?= "owner" ||
    @request.auth.projects_roles_via_user_id.roles_permissions_id.permission_ids.name ?= "roles.view"
  )
)
```

Depending on your actual use case and desired behavior, you may eventually need to reorganize your multiple `relations` fields into a more traditional SQL structure (_e.g. [junction table/collection](https://en.wikipedia.org/wiki/Associative_entity) that will have 2 single `relation` fields pointing back to the specific collections_) to avoid some of the caveats for the at-least-one vs match-all constraints.

If you haven't already, please refer also to the https://pocketbase.io/docs/api-rules-and-filters/ and https://pocketbase.io/docs/working-with-relations/ documentation. Additionally there should be also autocomplete in the Dashboard for the available rule fields.

**Reply by ganigeorgiev** [reply]
I'll stop engaging to avoid further confusion but if you need further help about the API rules, please export your collections configuration as json (Dashboard > Settings > Export collections) because again the rule will depend on the actual collection and field names and it is very difficult to provide any examples without being specific about you structure.

**Reply by lumosminima** [reply]
> And just to address the below because it bugs me:
> 
> > ... why market the framework as "Open Source backend for Your next SaaS". And yes this counts as marketing since their is time investment.
> 
> It is not marketing. I mentioned several times in the past and will repeat it again - **I'm not selling anything or promising anything more than what is already available.** If you consider the website text misleading, please suggest a better one, but behaving like somehow I own you something is wrong and you'll be further disappointed down the road if you approach PocketBase (or any other non-commercial MIT licensed project) with this attitude. Neither you, nor anyone else is entitled to anything. If you disagree, please use something else.

Even if you don't directly market it other people sill get influenced and promote it based on their initial impressions and promotion is considered marketing and I have nothing against it when the claims actually match reality. The claim is "Open source backend for your next SaaS app" but in reality you mentioned it several times that most of the feature set is for your other project presentator. The whole point of framework is to be somewhat general and taking into account variety of use cases and this is in conflict with current branding. So, the message could be anything but the current hero title as long as this project is tied to presentator because otherwise people are tricked into using this and submitting bug reports and feedback and you will only cherry pick usecases that support your own project and opinions rather than thinking of general use cases. 

As for homepage hero suggestion, you should not have one if you are not selling any thing or making any claim. You can check sqlite's homepage, it clearly tells what it is, what it will remain and I can check it's "when to use" page to see if it's for me. I'm totally cherry picking here but your own wording has a lot of conflict on faq:

> PocketBase was created to assist building self-contained applications that can run on a single server without requiring to install anything in addition (see [Presentator#183 ](https://github.com/presentator/presentator/issues/183)).

Pretty much anything can run on a single server and can become "self-contained" if you configure it that way. You can even run postgres embedded in go. So what does this mean exactly and which usescases are covered? Since my usecase of deriving a subset db is pretty common if you care that user should have ownership of their data (data takeout)


>Does it scale?

> Only on a single server, aka. vertical. Most of the time, you may not need the complexity of managing a fleet of machines and services just to run your backend.
> PocketBase could be a great choice for small and midsize applications - SaaS, mobile api backend, intranet, etc.
> Even without optimizations, PocketBase can easily serve 10 000+ persistent realtime connections on a cheap $4 Hetzner CAX11 VPS (2vCPU, 4GB RAM).
> You can find performance tests for various read&write operations in the official [benchmarks repo](https://github.com/pocketbase/benchmarks) .
> There is still room for improvements (I haven't done extensive profiling yet), but the current performance is already good enough for the type of applications PocketBase is intended for.

You say "I don't promise anything" but at same time say "It could be a great choice" . Also you say "I haven't done extensive profiling yet" but at same time say "current performance is good enough" how do you know that and what is your good enough?

> There are no promises for maintenance and support beyond what is already available (you can explore the [Roadmap](https://github.com/orgs/pocketbase/projects/2/views/1) to get a general idea where the project is headed but there are no fixed ETAs).

Please add this line upfront on homepage and mention there is only one person behind this project so people can better gauge if they want to use or not.

As for me, I wanted to learn golang and at same time build my SaaS app and this seemed like a perfect fit but I also counted on docs improvement with scenarios and examples to figure out most things every SaaS needs but that never happened and at this time I'm already using endpoints and even though your provided API rules work now it's still a bit cryptic to me and docs don't do a great job of explaining everything, I would have to ask similar guidance questions that may be "open ended" and don't want to waste your time and since I'm not heavily using the sdk mostly .send I'll look into something else if pocketbase doesn't work out for me. Thank you for your time!

**Reply by ganigeorgiev** [reply]
> You say "I don't promise anything" but at same time say "It could be a great choice" .

There is no promise in "it could be a great choice" unless you read "could" differently.

> Also you say "I haven't done extensive profiling yet" but at same time say "current performance is good enough" how do you know that and what is your good enough?

The "good enough" performance is based on the [linked benchmarks](https://github.com/pocketbase/benchmarks) above and the type of applications described in the above paragraph. Users will always have to determine on their own whether the presented benchmarks represents their use case or not.

In my opinion you are approaching open source projects wrongly because again neither me nor anyone else own you anything (even including information who created it or why the project exists). Nevertheless, I'll consider eventually clarifying it better but if you expect a warning on every page then this is just silly. If people don't want to spend time reading the documentation then I don't know how to assist further.

Every open source project is a piece of shared code that you have to decide whether you want to rely on it or not. **You can't demand any responsibility from a maintainer because the license itself explicitly rules this out.**

I think you are frustrated and trying to argue unnecessary but I really don't want to do that and I feel that whatever I say you'll remain disappointed., so kindly if you are not satisfied with PocketBase and how the project is managed, use something else.
